### PR TITLE
[IMP] base: add to automated actions the fields to watch during update

### PR DESCRIPTION
**Pull request by len-foss**
When users create an automated action to trigger when a record is updated,
they assume each update is 'atomic' in some sense.
It is not the case, as one write may trigger many write, some in recompute.
As a result, any update may trigger a dozen actions (or more).
We add a field wich allows us to specify which field we want to watch.
For instance, you could select the field 'state on quotation',
and thus only action would be triggered on confirming the quotation.

The link_field_id is a field that allows to link a record of the source model
to a new record of the target model, through a many2one.
However many target records can be created, so every one of them
(except the last) ends up orphan.
We allow for one2many and many2many fields to be used,
and in these cases we add the newly created record to the list.

opw 1910671

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/addons/base/models/ir_actions.py [hunk 1]
```diff
@@ -497,7 +497,10 @@ def run_action_object_create(self, action, eval_context=None):
 
         if action.link_field_id:
             record = self.env[action.model_id.model].browse(self._context.get('active_id'))
-            record.write({action.link_field_id.name: res.id})
+            if action.link_field_id.ttype == 'many2one':
+                record.write({action.link_field_id.name: res.id})
+            if action.link_field_id.ttype in ['one2many', 'many2many']:
+                record.write({action.link_field_id.name: [(4, res.id, True)]})
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_actions.py]: The command "4" is commonly a pair, not a triple
```suggestion
                record.write({action.link_field_id.name: [(4, res.id)]})
```

File: addons/base_automation/models/base_automation.py [hunk 2]
```diff
@@ -61,6 +61,9 @@ class BaseAutomation(models.Model):
     filter_domain = fields.Char(string='Apply on', help="If present, this condition must be satisfied before executing the action rule.")
     last_run = fields.Datetime(readonly=True, copy=False)
     on_change_fields = fields.Char(string="On Change Fields Trigger", help="Comma-separated list of field names that triggers the onchange.")
+    update_fields_id = fields.Many2many('ir.model.fields', string='Watched fields',
```
Comment (rco-odoo) [inline addons/base_automation/models/base_automation.py]: Naming convention says that the field should end with `_ids`.
I would also pick a more suggestive adjective, like `trigger_field_ids`.

File: addons/base_automation/models/base_automation.py [hunk 3]
```diff
@@ -193,8 +196,13 @@ def _process(self, records):
         # execute server actions
         if self.action_server_id:
             for record in records:
-                ctx = {'active_model': record._name, 'active_ids': record.ids, 'active_id': record.id}
-                self.action_server_id.with_context(**ctx).run()
+                # we process the action if we are in a create, or
+                # if we have update_fields_id, we check that we are writing on one of these fields
+                if not self._context.get('old_values') or not self.update_fields_id or (
+                        self._context.get('old_values') and
+                        set(self._context.get('old_values')[record.id]) & set(self.update_fields_id.mapped('name'))):
```
Comment (rco-odoo) [inline addons/base_automation/models/base_automation.py]: I find the condition a bit hard to read. What about this?
```python
old_vals = self._context.get('old_values', {}).get(record.id)
if not old_vals or not self.update_fields_id or any(
    field.name in old_vals for field in self.update_fields_id
):
```
The `any(...)` part translates easily as "some of the trigger fields has been updated."
(Note: this is subjective and reflects personal taste.)

**Comment by nim-odoo** [comment]
> "If it’s not tested, it’s broken." (Bruce Eckel)
> Tests please!

"Tester, c'est douter." (Antony Lesuisse)

File: addons/base_automation/models/base_automation.py (hunk 2 above)
Comment (len-odoo) [reply addons/base_automation/models/base_automation.py]: Good suggestion, thanks!

File: addons/base_automation/models/base_automation.py (hunk 3 above)
Comment (len-odoo) [reply addons/base_automation/models/base_automation.py]: It's a much better rewriting. Note that this condition was still false, and I moved it to a function to declutter that part of the code.

**Comment by len-odoo** [comment]
Naming of _trigger_field is not good, any idea? 
A test has been added.

File: addons/base_automation/models/base_automation.py [hunk 4]
```diff
@@ -61,6 +61,9 @@ class BaseAutomation(models.Model):
     filter_domain = fields.Char(string='Apply on', help="If present, this condition must be satisfied before executing the action rule.")
     last_run = fields.Datetime(readonly=True, copy=False)
     on_change_fields = fields.Char(string="On Change Fields Trigger", help="Comma-separated list of field names that triggers the onchange.")
+    trigger_field_ids = fields.Many2many('ir.model.fields', string='Watched fields',
```
Comment (rco-odoo) [inline addons/base_automation/models/base_automation.py]: I was wondering whether we could merge `trigger_field_ids` and `on_change_fields`.
They are both used in a similar way, except for the empty value: it always triggers the action for create/write, while it never triggers it for onchange. But an empty list in the case of onchange does not make sense...

File: odoo/addons/base/models/ir_actions.py [hunk 5]
```diff
@@ -497,7 +497,10 @@ def run_action_object_create(self, action, eval_context=None):
 
         if action.link_field_id:
             record = self.env[action.model_id.model].browse(self._context.get('active_id'))
-            record.write({action.link_field_id.name: res.id})
+            if action.link_field_id.ttype == 'many2one':
+                record.write({action.link_field_id.name: res.id})
+            if action.link_field_id.ttype in ['one2many', 'many2many']:
+                record.write({action.link_field_id.name: [(4, res.id)]})
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_actions.py]: This fixes another issue, doesn't it?

Comment (rco-odoo) [reply odoo/addons/base/models/ir_actions.py]: made a second commit

**Comment by len-odoo** [comment]
You directly made the changes then.
Thanks a lot.

robodoo r+

**Comment by rco-odoo** [comment]
robodoo r-

**Comment by rco-odoo** [comment]
robodoo rebase-ff r+

**Pull request closed without merging**